cordformation/src/main/kotlin/net/corda/plugins/Dockerform.kt

```diff
@@ -146,33 +144,70 @@ open class Dockerform @Inject constructor(objects: ObjectFactory) : Baseform(obj
 
                 it.installDefaultDatabaseConfig(dbConfig)
 
-                val args = mutableMapOf<String, Any>()
-
-                dockerfileArgs.entrySet().toList().map {
-                    args[it.key] = it.value.unwrapped()
-                }
-
-                services[dbHost] = mapOf(
+                val dbDockerfile = dockerConfig.getString("dockerConfig.dbDockerfile")
+                // Override the port and hostname parameters in dockerfile
+                val dbDockerfileArgs = defaultUrlArgs.withFallback(
+                        dockerConfig.getConfig("dockerConfig.dbDockerfileArgs")).resolve()
+                val database = mutableMapOf(
                         "build" to mapOf(
                              "context" to project.buildDir.absolutePath,
                              "dockerfile" to project.buildDir.resolve(dbDockerfile).toString(),
-                             "args" to args
+                             "args" to dbDockerfileArgs.entrySet().map { it.key to it.value.unwrapped() }.toMap()
                         ),
-                        "restart" to "unless-stopped",
-                        "ports" to listOf(dbPort)
+                        "restart" to "unless-stopped"
                 )
 
+                // append persistence volume if it is required
+                if (dockerConfig.hasPath("dockerConfig.dbDataVolume")) {
+                    var hostPath = dockerConfig.getString("dockerConfig.dbDataVolume.hostPath")
+                    var containerPath = dockerConfig.getString("dockerConfig.dbDataVolume.containerPath")
+
+                    if (dockerConfig.hasPath("dockerConfig.dbDataVolume.pathArgs")) {
+                        val pathArgs = dockerConfig.getConfig("dockerConfig.dbDataVolume.pathArgs")
+
+                        hostPath = ConfigFactory.parseString("HOSTPATH=${TypesafeUtils.encodeString(hostPath)}")
+                                .resolveWith(ConfigFactory.empty().withValue("NODENAME", ConfigValueFactory.fromAnyRef(it.nodeDir.name)))
```

Comment:
This `NODENAME` substitution variable is bothering me. As I see it, `hostPath` is either a relative path or an absolute one. And if it's relative then should be relative to `nodeDir`. In other words, I would expect something like:
```kotlin
var hostPathStr = dockerConfig.getString("dockerConfig.dbDataVolume.hostPath")

// etc...

hostPathStr = ConfigFactory.parseString("HOSTPATH=${TypesafeUtils.encodeString(hostPathStr)}")
    .resolveWith(pathArgs)
    .getString("HOSTPATH")

// etc...

val hostPath = Paths.get(hostPathStr)
val absoluteHostPath = if (hostPath.isAbsolute) {
    hostPath
} else {
    nodeBuildPath.resolve(hostPath)
}
```
I realise that some of these objects are `String` and others are `Path`, but I hope you get the general idea. To be honest, a `Path` object is a _lot_ more powerful than a `String`, so I'd be inclined to keep hold of the node's directory, e.g.:
```kotlin
val nodeBuildPath = directoryPath.resolve(it.nodeDir.name).toAbsolutePath()
val nodeBuildDir = nodeBuildPath.toString()
```
I like the reuse of the `TypesafeUtils.encodeString()` function. Looking at your `dbDataVolume` elements from the _outside_, I see two elements called "host**Path**" and "container**Path**", and a map called "**path**Args". So why _wouldn't_ I expect the substitutions to be applied to _both_ elements equally?